src/Compilers/Core/Portable/Symbols/ISymbolExtensions.cs

```diff
@@ -116,5 +118,83 @@ internal static bool IsInSource(this ISymbol symbol)
 
             return false;
         }
+
+        // Keep consistent with TypeSymbolExtensions.IsWellKnownTypeLock.
+        internal static bool IsWellKnownTypeLock(this ITypeSymbol type)
+        {
+            return type is INamedTypeSymbol
+            {
+                Name: WellKnownMemberNames.LockTypeName,
+                Arity: 0,
+                ContainingType: null,
+                ContainingNamespace:
+                {
+                    Name: nameof(System.Threading),
+                    ContainingNamespace:
+                    {
+                        Name: nameof(System),
+                        ContainingNamespace.IsGlobalNamespace: true,
+                    }
+                }
+            };
+        }
+
+        // Keep consistent with LockBinder.TryFindLockTypeInfo.
+        internal static LockTypeInfo? TryFindLockTypeInfo(this ITypeSymbol lockType)
+        {
+            IMethodSymbol? enterLockScopeMethod = TryFindPublicVoidParameterlessMethod(lockType, WellKnownMemberNames.EnterLockScopeMethodName);
+            if (enterLockScopeMethod is not { ReturnsVoid: false, RefKind: RefKind.None })
+            {
+                return null;
+            }
+
+            ITypeSymbol? scopeType = enterLockScopeMethod.ReturnType;
+            if (!(scopeType is INamedTypeSymbol { Name: WellKnownMemberNames.LockScopeTypeName, Arity: 0, IsValueType: true, IsRefLikeType: true, DeclaredAccessibility: Accessibility.Public } &&
+                lockType.Equals(scopeType.ContainingType, SymbolEqualityComparer.ConsiderEverything)))
```

Comment:
Please either do the DeMorgan's expansion, or indent the second line of this `if` so it doesn't imply that the `Equals` call is outside the parens.
```suggestion
            if (scopeType is not INamedTypeSymbol { Name: WellKnownMemberNames.LockScopeTypeName, Arity: 0, IsValueType: true, IsRefLikeType: true, DeclaredAccessibility: Accessibility.Public } ||
                !lockType.Equals(scopeType.ContainingType, SymbolEqualityComparer.ConsiderEverything))
```